controlcenter: many setting panes and minor features

atdma opened the pull request:
I have added the following panels in the controlcenter.

- Ethernet
- Wireless
- Audio
- Appearance
- Taskbar

They can continue to be built upon easily, I tried to match the original style that was left in the work in progress panels.

I have also added an icon to the system tray for ethernet connections and a popout for this icon.

andrsrjld commented:
look like we have working on the same feature 😀

atdma commented:
I've added a bit more, currently with mind on a mission to have the controlcenter able to manage most aspects of the system.

atdma commented:
https://github.com/user-attachments/assets/850959e0-a635-4bb8-85b1-d1250d1f93d5

A little preview of whats been accomplished.

atdma commented:
I have successfully reworked the toasts and notification system. The following changes can be seen.

- Toasts still work as expected.
- Notifications on a notif event will be shown as a toast in the top right corner.
- Notifications panel added, will popout when in top right corner.
- Notifications and Lockscreen now use the same list of notifications.
- Clear all button for notifications, will also clear the notifications on the lockscreen.

https://github.com/user-attachments/assets/7ef30b7d-0087-4d08-9000-9e4cc206bfd6

Justsnoopy30 commented:
I feel like having the notifications float like that is a bit out of place with the shell to be honest, as everything else extends out from the outline around the screen and feels very integrated into it. It would make more sense for that to be how notifications look in fullscreen mode, though currently the shell doesn't render anything in fullscreen and making it do so when receiving a notification could cause a lag spike as direct scanout has to be turned off for the duration of the overlay.

Though, the notification panel itself isn't a bad idea. The quick clear button being outside of it has the same issue of not being attached to the shell but not sure it would look good inside...

soramanew commented:
Very nice work, I really like it! One thing though, perhaps the wireless and ethernet settings should be combined into a single pane named network or something, cause they aren't really separate.

nuknes commented:
> Feeling pretty good about these revisions, ended up writing an entire nmcli wrapper to make the networking panels in the controlcenter much more reliable. Have been polishing the UI of the controlcenter and now looks better than the original video above. Everything has been split out into components when needed to keep the code clean, maintainable, extensible. Ready for more features to be added as people want to grow the project more.
> 
> Check it out, thanks everyone that has given feedback so far!

I've been following your non stop marathon these few days and it's been a great learning experience watching you drop in flesh out and refine the controlcentre. looks great and can't wait for the final polish and merge. thank you and go get some rest lol

atdma commented:
https://github.com/user-attachments/assets/acdc803c-1068-4edc-81f3-2dbc50b8cf78

I've accomplished all of the feature requests that have been made in this thread.

Currently contemplating ways to integrate the password input dialog into the popout for wireless connections in the tray of the bar.

atdma commented:
https://github.com/user-attachments/assets/21cfcba3-029a-4eea-966d-7fc9714c2222

Alright, my goal has been completed. I hope everyone likes the additions.

Justsnoopy30 commented:
Have you considered adding a search bar? The launcher page could really use one to find apps. A global search could be useful too (though less so due to the good categorization and limited amount of settings).

atdma commented:
> Have you considered adding a search bar? The launcher page could really use one to find apps. A global search could be useful too (though less so due to the good categorization and limited amount of settings).

I did make a feature like that at one time. Let me dig through my dump spot on my NAS and see if I can build that functionality up again. I could see other people liking that as well. If it works well I'll push the commits.

atdma commented:
@PixelKhaos I did get per-application audio sliders working in the controlcenter, with uv type level monitoring on the main output. I have not found a way to get the uv monitoring per-application with only qml. It seems I can only do it with the default sink.

I'm gonna wait to push any audio specific stuff like that because I'll likely build it into a few different panels/popout. I would like to spend extra amount of time making it beautiful. If this gets merged, I'll start a new branch on my repo for specific features so I can work on multiple things without breaking anything.

atdma commented:
@Justsnoopy30 I started running into walls when attempting to make it a global search. I did the next best thing though.

https://github.com/user-attachments/assets/457ce824-3cd2-4dee-8c1a-fbfa8df96aa4

soramanew commented:
> Might be best honestly to polish what you already have instead of scope-creeping this PR. Additional functionality should be added later

I agree, this PR is already extremely massive. I appreciate all the work done in it, however having such a massive pr makes it pretty hard to review. Next time, please split it into multiple prs thanks.

PixelKhaos commented:
> Which other aspects do you think need polished. This has been polished a bit more since the videos. I would recommend running it and seeing what you think. I'm definitely cool with cleaning up anything you think may be off.

Was more of a general statement, with this many rapid changes, its been hard to keep an eye on it all to know what's relevant or not anymore. Essentially, if you feel like its good to go where it's at now, just wait for reviews and dont push new additions or changes other than those reviews

soramanew commented:
@atdma a few issues I found without looking at the code:

<img width="259" height="56" alt="image" src="https://github.com/user-attachments/assets/a3503a16-fac9-40e1-8505-739f3e57a773" />

The icon is too big, too bold and has too much spacing.

<img width="657" height="338" alt="image" src="https://github.com/user-attachments/assets/7b826c34-eea1-42bf-b655-7ecf7531d490" />

When you open one of the sections, it automatically collapses any other section. This behaviour is kinda unexpected imo, especially on the network and audio panes where there are only 2 sections.

<img width="1639" height="487" alt="image" src="https://github.com/user-attachments/assets/ffc0fbd8-cc51-4329-9543-87aafd8ca045" />

Ahem...

<img width="682" height="43" alt="image" src="https://github.com/user-attachments/assets/16d47a57-a06f-4d0a-8c06-7480c9644406" />

The content isn't clipped so when transparency is enabled, you can see behind the borders.
Also, the network pane doesn't have a scrollbar.

<img width="1010" height="995" alt="image" src="https://github.com/user-attachments/assets/4686e36f-51d7-4de0-8b32-0433a2c1825c" />

The audio pane gets stretched if the window is too tall.

<img width="666" height="325" alt="image" src="https://github.com/user-attachments/assets/06e244ac-befe-44f3-a55b-e8453de3c0e1" />

Try not to use outlines too much, and especially not primary outlines. Only use outlines with outline buttons (which use a outline colour for the border) or for special states (e.g. the password input when it has focus).

<img width="395" height="137" alt="image" src="https://github.com/user-attachments/assets/cd7e1282-d2e5-489a-b59e-8b94529d73f5" />

The ripple isn't rounded to the shape of the password box. Also, don't use full rounding for the buttons here.

<img width="753" height="645" alt="image" src="https://github.com/user-attachments/assets/93c1c80f-f875-45a0-b6ae-4b66c5540578" />

This looks kinda empty, also is there supposed to be something below the taskbar entries bit?

<img width="978" height="568" alt="image" src="https://github.com/user-attachments/assets/e37c3c44-f611-438e-b69b-3f4993282849" />

This also looks empty, also that toggle doesn't make sense when nothing is selected.

If it's alright with you, I can directly fix some stuff on your branch.

soramanew commented:
Also, I think a cursor for the password input would be better than an outline. It just feels kinda weird without one when you're supposed to type into it.

soramanew commented:
And no need to spam the logs with this

<img width="706" height="994" alt="image" src="https://github.com/user-attachments/assets/32205568-3e1b-451d-af57-adf748b8d878" />

Justsnoopy30 commented on modules/background/Background.qml:
```diff
@@ -9,7 +9,6 @@ import Quickshell.Wayland
 import QtQuick
 
 Loader {
-    asynchronous: true
```
Why was this necessary to be changed with this PR? I recall seeing in one of the issues that doing this caused issues (https://github.com/caelestia-dots/shell/pull/847#issuecomment-3451783973, https://github.com/caelestia-dots/shell/pull/847#issuecomment-3452503116), and there was a different fix applied in Wallpaper.qml a while back.

atdma replied:
I appreciate you bringing that up, I was acknowledging this as a bug that would cause the wallpaper to not load. If there are other reasons for it being there, we can put it back.

nuknes replied:
@atdma 
`asynchronous: true` here isn't a bug and wasn't causing the wallpaper to not load. The wallpaper issue is already fixed in Wallpaper.qml

If this was made a synchronous execution, the file i/o involved in starting the wallpaper image load blocks the main thread and delayed other daemon/ service startups creating performance bottleneck/ timeouts

atdma commented:
@soramanew the cursor blink and the outline on focus in the password input popout are being elusive the way that it is written, I am going to sidebar these for now. The other corrections for the password popout have been corrected. I'm going to move on to the rest of the list.

atdma commented:
@Justsnoopy30 I have added the signal strength icons in the wireless network list to match the popout from the tray.

I also have done a slight refresh on the audio pane to make it a bit more to the material 3 guidelines.

atdma commented:
https://github.com/user-attachments/assets/85d52df8-09f4-4d20-b711-40dcfaa18b8c

Latest revision for anyone that wanted a quick peek.

Justsnoopy30 commented:
Dunno if it's easily doable (or would fit the ui well, can't tell without trying it), but if you were able to layer the lock icon onto the wifi signal icon, it would also get rid of the need to write secured and stand out more at a glance for people looking to press the closest open wifi on the go. It would also make it match with the popout better, which uses an icon. Other than that, great work!

soramanew commented:
Btw @atdma I don't think `Qt.callLater` works how you're expecting it to.

<img width="475" height="118" alt="image" src="https://github.com/user-attachments/assets/5e8138c6-c421-4dd6-8c70-fc50321ba6ed" />

This does not run the function after a timeout of 150ms, `Qt.callLater` runs the function after the QML engine returns to the event loop, i.e. on the next event loop iteration.

Here are the docs https://doc.qt.io/qt-6/qml-qtqml-qt.html#callLater-method

soramanew commented:
Use the Config singleton instead please

atdma commented:
> Use the Config singleton instead please

Got it, I can rewrite that if you want.

Justsnoopy30 commented:
Checked out locally - noticing that the chosen wallpaper flickers and reloads the thumbnail when switched in the settings app.

atdma commented:
@Justsnoopy30 @soramanew I believe I've gotta tap out for the night, I'm getting sleepy and gonna mess stuff up after fixing. Thanks for attacking this with me and dealing with my mistakes, I'm not a quickshell pro yet.

Those have steps/intervals that can be set, when I set those up I didn't know what a good range would be; it borderline hasn't been touched since I created the button.

Last things on my list I haven't gotten to was the focus on password popout, and re-fixing the edge-to-edge clip in the panes.

Have a good night guys.

atdma commented:
> recording_20251115_15-30-49.mp4
> 
> The background tone behind sliders does not show up when transparency is disabled.
> 
> ~Also, I see that backgrounds show up behind the top bar when transparency is on, looking really weird.~ (assuming this is covered under "Edge to edge corrections that can be viewed with transparency.")

Thanks I'll add that to the list.

atdma commented:
Sounds good, thanks for narrowing down the priorities and understanding. I'll focus on some of that now and will update this thread with progress.

atdma commented:
I believe all of those features have been completed.

- There could be an edge case I may have missed in the odd content disappearing during animation; but I believe I fixed it.
- I added a safe way for the config singleton to save config, so saving options is now working as expected.
- Converted all numerical settings to sliders and adjusted min-max, stepping intervals, etc.
- Added the ability to manually type in values for the numerical sliders.
- I was not able to replicate the background disappearing on the sliders with transparency disabled.
- Applications now fade in, with the same effect and duration as the wallpapers.

Justsnoopy30 commented:
Oh so we're thinking of different things, I'm talking about the tone indicating what part the slider is not covering. You can see in your video that it's gone.
with transparency:
<img width="438" height="99" alt="image" src="https://github.com/user-attachments/assets/29770af5-9082-49d2-94b2-bb544f4b29d7" />

without:
<img width="438" height="94" alt="image" src="https://github.com/user-attachments/assets/8bfe3c93-e659-4cb8-9463-efb9ee815175" />

(Github literally did not show me your message until now...)

Justsnoopy30 commented:
Also we should really uh, tweak the padding slider maximum lol.
<img width="1918" height="1078" alt="image" src="https://github.com/user-attachments/assets/fe5d404c-0906-4597-b587-9ebab143a1d4" />

Kinda impossible to get back without changing the config file.

The padding slider is also kind of broken because the live padding change messes with the mouse positions and jerks around the slider every tiny movement. But it's not much of an issue if you just make the maximum much smaller.

Justsnoopy30 commented:
> There could be an edge case I may have missed in the odd content disappearing during animation; but I believe I fixed it.

Still seeing this btw, not sure what changed with that.

https://github.com/user-attachments/assets/0a550071-4533-4834-b7a2-16bbb82a13d7

atdma commented:
> > There could be an edge case I may have missed in the odd content disappearing during animation; but I believe I fixed it.
> 
> Still seeing this btw, not sure what changed with that.
> recording_20251115_20-03-05.mp4

I must have misunderstood what you were describing, this was much better to see it.

atdma commented:
> Border curve around wallpaper at bottom half is getting filled with black instead of being transparent

Let me know if that resolved all of the issues, I believe everything should be good.

Justsnoopy30 commented:
After something you just changed the font subsection is uh, well, just click it.

Justsnoopy30 commented:
Oh are you not getting it? The font section takes like 5 seconds to open. I'm getting this:

https://github.com/user-attachments/assets/7959bd75-ab8f-4c4f-8367-d2a3436799e1

atdma commented:
> Oh are you not getting it? The font section takes like 5 seconds to open. I'm getting this:
> recording_20251115_20-34-10.mp4

You may have a ton more fonts than I, I'm really unsure but mine was very responsive.

Personally I think we should remove the fonts section, generally it will ruin the aesthetics of peoples setup by changing the fonts, and if they really want to change the fonts, they could always edit json.

Justsnoopy30 commented:
Selected wallpaper outline is cut off at the bottom
<img width="214" height="160" alt="image" src="https://github.com/user-attachments/assets/e7822602-b230-484f-85a4-92ecf75804ee" />

I'm scared to bring this up cause you might go insane but there appears to be a one-pixel off-by-one issue with the bottom 2 corners of wallpapers causing white pixels... wouldn't bother if you have no idea what does that. It might get fixed with the outline fix.
<img width="29" height="148" alt="image" src="https://github.com/user-attachments/assets/0f0e12aa-b0f8-42e3-8220-d13b39ae90df" />
<img width="421" height="52" alt="image" src="https://github.com/user-attachments/assets/33c762d0-4450-4f54-82c0-814750fd3805" />

Justsnoopy30 commented:
It looks like the opacity of the text is not getting faded like the other items, just getting clipped

atdma commented:
> It looks like the opacity of the text is not getting faded like the other items, just getting clipped

I believe I know what it is, give me 5-10 minutes there is some deleting involved.

atdma commented:
5-10 minutes, turned into over an hour. It should be better now.

![20251115234630](https://github.com/user-attachments/assets/5e6b73d8-be63-4b21-bf14-a6857c75124c)

Justsnoopy30 reviewed (commented):
I don't see anything else besides the issues in the comment above that are still remaining and the known things in the megalist that are left for future contributions.

atdma commented:
> Well I guess removing the ability to popout is one way to workaround the resizing issue by making it non-resizable, but I think having a button to toggle it into a window is still very useful even if you just disable resizing on it so that people can move it to a second monitor or out of the way without losing their place in the settings menu.
> 
> It wasn't an issue when it was much smaller, but now that there's a lot more settings I find myself constantly having to close and reopen the control center to do anything else like comment on this thread whereas before I could keep reporting stuff without reopening over and over.

That is how the other popouts behaved, I figured making it match would knock out two things at once. I could restore the float and disable resize. No problemo.

Justsnoopy30 commented:
Are you seeing this on your end? (Animation broken and flicker on dashboard now)

https://github.com/user-attachments/assets/361d0cfc-0eaf-4c24-9e31-a4cbc01886c2

atdma commented:
> Nope, still happening. Occurs with [465df12](https://github.com/caelestia-dots/shell/commit/465df12855886f82ea2259aee5b7fd368bdbd129) apparently? Reverting that fixes it.

Thank you for narrowing it down, I'll revert to that commit and update my log above of fixes.

Justsnoopy30 commented:
I meant reverting that commit fixes it, not reverting TO that commit. It's the utilities circular import correction commit that broke it, not sure what exactly that is doing. It's still broken in what you pushed. No idea why you're not seeing it on your end.

atdma commented:
> I meant reverting that commit fixes it, not reverting TO that commit. It's the utilities circular import correction commit that broke it, not sure what exactly that is doing. It's still broken in what you pushed. No idea why you're not seeing it on your end.

I'll go back one more, I'm attempting to find a stable state on both of our systems to grow from.

atdma commented:
> > @Justsnoopy30 I'm contemplating options, I really enjoy how smooth it runs as a popout. I also see the side of its easier to use as a floating window. There is something going on with async/sync as a float vs popout.
> 
> > Any ideas are welcome.
> 
> I don't know much about QT internals tbh so I wouldn't know why that would be the case, but we can ask soramaine when they review again maybe? For now I think you could just make both cases a popout by default, fix the animation, and add back a button to manually switch to floating mode for testing.

I like the suggestion, I'll do this and leave a note where I'm at.

Justsnoopy30 commented:
Okay the issue is no longer present now. I wonder if maybe you missed running a rebuild if any of the commits you made touched the files that needed it to take full effect? I run 
```                    
cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/
cmake --build build
sudo cmake --install build
``` 
after each pull to make sure.

atdma commented:
Currently just gotta work out that animation issue on show/hide as a popout, I have restored the rest of the code to the working state.

atdma commented:
> Minor but still here lol in theme mode, color scheme, color variant. As you can see, it's getting clipped instead of faded. Works everywhere else like the audio tab and such.

This problem should be resolved, currently I am at my wits end, each attempt for me to fix the animation on open and close has failed. I'm going to take a break for the moment.

Some of the other things such as the float vs window may have to wait for @soramanew to look.

StPlay1 commented:
Hey everyone, I used Celestia Shell early in my Arch journey (when I didn't understand much about it, rice, and things like that). Now that I've looked into a few things, I'd like to ask how to add these panela, and o want to know If It os oficial.

atdma commented:
@Justsnoopy30 I should've been smart and turned my animation scale to 5x. :D I appreciate your patience, with my latest revision (not pushed this second) I have resolved this issue.

This revision seems to solve it once and for all, I hope. 🙏

https://github.com/user-attachments/assets/bc80848a-3317-4aa8-9346-d6cedddd17ab

Justsnoopy30 commented:
Awesome! I think the only thing left once that is pushed then would be preventing stutter on the opening animation? It looks like you ended up doing almost everything on the mega list after all lol

Justsnoopy30 commented:
> Right. You said that Isnt ready to use, but there is a way to test It? Im in my journey trying to learn quickshell and o think that i can learn, and solve my problema..

I mean yes you can checkout the branch and build it locally (see readme except you are cloning the atdma repo instead)... but it's almost done anyways and shouldn't be too long before it's in I think. There's been plenty of testing.

StPlay1 commented:
> > Right. You said that Isnt ready to use, but there is a way to test It? Im in my journey trying to learn quickshell and o think that i can learn, and solve my problema..
> 
> I mean yes you can checkout the branch and build it locally (see readme except you are cloning the atdma repo instead)... but it's almost done anyways and shouldn't be too long before it's in I think. There's been plenty of testing.

   Sorry, one last thing. How i do It? I I don't know where the files are usually located... Every little tutorial helps!

Justsnoopy30 commented:
> controlcenter: removed accordion auto-close on sections

@soramanew can you clarify if you thought this should be removed for every section or just the ones you mentioned with few tabs?

Justsnoopy30 commented:
> Sorry, one last thing. How i do It? I I don't know where the files are usually located... Every little tutorial helps!

Follow https://github.com/caelestia-dots/shell?tab=readme-ov-file#manual-installation
but instead of the line `git clone https://github.com/caelestia-dots/shell.git caelestia`, use
`git clone https://github.com/atdma/shell.git caelestia`

Justsnoopy30 commented:
Okay so issue, wallpapers are being loaded synchronously and not being lazy loaded either apparently. I managed to completely lock up the shell and had to force kill the process after adding a bunch of wallpapers and opening the control center.

atdma commented:
> Okay so issue, wallpapers are being loaded synchronously and not being lazy loaded either apparently. I managed to completely lock up the shell and had to force kill the process after adding a bunch of wallpapers and opening the control center.

I'll look into this shortly.

Justsnoopy30 commented:
Ah whoops I missed that you mentioned the animation issue in your message. I can give sora a ping on discord about doing another review as they probably stopped checking this thread after getting over a hundred emails...
Should I wait for you to try yourself?

Justsnoopy30 commented:
> I have tried a few times, I've had very little success on that particular error. Even if sora just popped in and highlighted some relevant code to help the situation I may be able to get it done.

Gave sora a ping btw, they're just busy rn but they still check it.

soramanew commented:
More stuff:

use the lock variants of the network icons here instead of a separate lock icon.
also move the connected state and security to beneath the ssid the same way the bt pane does it, the spacing looks too large without it
<img width="631" height="532" alt="image" src="https://github.com/user-attachments/assets/ccd9a48d-51b0-448a-9fad-f7f2a83f8462" />

scrolling the right side of the network panel when a wifi network is selected doesn't work

Justsnoopy30 commented:
> scrolling the right side of the network panel when a wifi network is selected doesn't work

to add: same behavior with bluetooth panel, never noticed this ig

Justsnoopy30 commented:
No spacing between the tap-animation and the contents below looks pretty weird
<img width="431" height="134" alt="image" src="https://github.com/user-attachments/assets/2d58109f-4621-409d-901e-e4ffe50f1d4e" />

Shade of circle containing lock icon on wifi looks a bit odd for the selected wifi, it blends too much into the background color imo
<img width="71" height="66" alt="image" src="https://github.com/user-attachments/assets/b953dfc6-71c9-49ba-bcf0-96e80cd60b53" />

The new taskbar section feels much better. It could use a little bit more spacing between elements and between Settings text and the first element, though, I think. Maybe try matching it with the spacing you see between elements in the dashboard and utilities panel, which both have more? Not sure.

Also, I think it could be even "caelestia" feeling if we did a gallary view for the appearance section too (and more helpful as you could see all the options at once), but I'm not sure where the wallpaper section could go then. Maybe something to prototype?

Unrelated to GUI, but just noticed because the button is exposed in the UI: Caps lock status icon doesn't show up when that setting is on unless the shell starts with caps lock on. It needs to show/hide based on the live state of caps lock.

Edge case but did want to mention, stuttering issue is not fully fixed, you can still encounter it if you switch tabs too quickly before the gui has had time to load stuff right after it finishes the opening animation. Don't know what you could do about that at the moment as it sounded like you made everything you could async, but something to consider if you have other ideas. Just wanted to mention as you checked it off the mega list, it's only much less severe but not gone in every scenario.

Another edge case: If you manage to scroll tabs before the opening animation is complete (can happen at animation speed 1 but more prominent at slower scales), pages go blank. 

Both of these could actually be real encountered issues that will pop up frequently if/when a keybind to open the settings menu is added, as your mouse will already be sitting right where the settings window is, and it's feasible that people will start scrolling immediately and encounter blank pages or stuttering. Again, amplified on slower computers or power saving mode.

soramanew commented:
Wdym by popout initiated control centres?

atdma commented:
@soramanew

Also after looking into the WallpaperGrid.qml that I am using in the controlcenter wallpaper selector, I can confirm that it is using the service, utilizing the methods Wallpapers.list, Wallpapers.setWallpaper, and Wallpapers.actualCurrent.

I removed a manual filename extraction because it already provided a name property, this seems to be using the single source of truth as the wallpaper service.

Chujo58 commented:
That looks amazing! Really hyped to see this merged!
